### PR TITLE
Add the declaration engine skeleton code

### PR DESCRIPTION
**Pull request by emilyaherbert**
This PR adds the skeleton code for the declaration engine.

### CODE REVIEW

**Review by otrho — APPROVED** [review]

File: sway-core/src/declaration_engine/declaration_engine.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    pub(crate) fn look_up_decl_id(&self, index: DeclarationId) -> DeclarationWrapper {
+        self.slab.get(index)
+    }
+
+    pub(crate) fn add_monomorphized_copy(
+        &mut self,
+        original_id: DeclarationId,
+        new_id: DeclarationId,
+    ) {
+        match self.monomorphized_copies.get_mut(&*original_id) {
```
Comment (vaivaswatha) [inline sway-core/src/declaration_engine/declaration_engine.rs:36]: Using `HashMap::entry()` could be more efficient here.

File: sway-core/src/declaration_engine/declaration_engine.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+};
+
+use super::{declaration_id::DeclarationId, declaration_wrapper::DeclarationWrapper};
+
+/// Used inside of type inference to store declarations.
+pub struct DeclarationEngine {
+    slab: ConcurrentSlab<DeclarationId, DeclarationWrapper>,
+    // *declaration_id -> vec of monomorphized copies
+    // where the declaration_id is the original declaration
+    monomorphized_copies: im::HashMap<usize, Vec<DeclarationId>>,
```
Comment (vaivaswatha) [inline sway-core/src/declaration_engine/declaration_engine.rs:16]: Any reason to not use `HashMap<DeclarationId, ...>`?

**Comment by emilyaherbert** [comment]
> @vaivaswatha I have some changes coming up that deal with that code so approving this now, will take your feedback into account. 

Not sure I know what you mean by this @vaivaswatha ?

File: sway-core/src/declaration_engine/declaration_engine.rs (hunk 2 above)
Comment (emilyaherbert) [reply sway-core/src/declaration_engine/declaration_engine.rs:16]: Just a regular HashMap would probably be better.

**Pull request merged**